### PR TITLE
Subtype FRU fields first in `type_changing_struct_update`

### PR DESCRIPTION
**Pull request by compiler-errors**
So this fixes a subtle bug that `type_changing_struct_update` introduced, where it'll no longer coerce the base expr correctly. I actually think this code is easier to understand now, too.

r? @lcnr since you reviewed the last one

### CODE REVIEW

File: compiler/rustc_typeck/src/check/expr.rs [hunk 1]
```diff
@@ -1561,71 +1561,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                                    }
+                                    // FIXME: Need better diagnostics for `FieldMisMatch` error
+                                    Err(_) => {
+                                        self.report_mismatched_types(
+                                            &cause,
+                                            target_ty,
+                                            fru_ty,
+                                            FieldMisMatch(variant.name, ident.name),
+                                        )
+                                        .emit();
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/expr.rs:1595]: I wonder if we should return `TyKind::Error` here

Comment (lcnr) [reply compiler/rustc_typeck/src/check/expr.rs:1595]: probably won't hurt, but maybe it's better to wait until we get a test case for where that's needed? :thinking:

File: compiler/rustc_typeck/src/check/expr.rs [hunk 2]
```diff
@@ -1561,71 +1561,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                                            )
-                                            .emit();
-                                        }
-                                    }
-                                }
-                                self.resolve_vars_if_possible(fru_ty)
-                            })
-                            .collect()
+                    if !error_happened && !base_ty.references_error() {
+                        fru_tys
```
Comment (lcnr) [inline compiler/rustc_typeck/src/check/expr.rs:1629]: why is that needed? What happens if we always use `fru_tys` here

**Review by lcnr — COMMENTED** [review]
so to summarize:

The current approach first does subtyping + opt coercion from the base expr to `TargetTy<?0..?N>` where `?0` to `?N` are fresh inference var.

The second approach first does subtyping of `TargetTy<?0..?N>` to our actual target ty and then does subtyping + coercions between the inferred target ty and the base type.

The new approach can do more coercions as we know more about the target type at that point.

File: compiler/rustc_typeck/src/check/expr.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/expr.rs]: yeah, it's way too overprotective i think

File: compiler/rustc_typeck/src/check/expr.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/expr.rs:1596]: yeah true
